From 2997342dbf8711f4dcf89c575cfce5ca95fdc7b9 Mon Sep 17 00:00:00 2001
From: Tom Lukaszewicz <tluk@chromium.org>
Date: Fri, 15 Jul 2022 20:38:39 +0000
Subject: [PATCH 093/110] [M102 Merge] Fix UAF in
 CloseBubbleOnTabActivationHelper

There are circumstances where a bubble may outlive its hosting
browser window (e.g. the browser is closed before the attached
bubble).

This CL adds an observation on the BrowserList to nullify the
Browser member if the Browser is destroyed before the activation
helper.

(cherry picked from commit 7c209778702c4f291c6d3b1fb9ac3cc803e0d90a)

Bug: 1341603
Change-Id: Ife3a5072b7009f8b42c6c68098c0245798384451
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3753347
Commit-Queue: Thomas Lukaszewicz <tluk@chromium.org>
Reviewed-by: Elly Fong-Jones <ellyjones@chromium.org>
Cr-Original-Commit-Position: refs/heads/main@{#1022874}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3766064
Owners-Override: Prudhvikumar Bommana <pbommana@google.com>
Bot-Commit: Rubber Stamper <rubber-stamper@appspot.gserviceaccount.com>
Commit-Queue: Srinivas Sista <srinivassista@chromium.org>
Cr-Commit-Position: refs/branch-heads/5005@{#1252}
Cr-Branched-From: 5b4d9450fee01f821b6400e947b3839727643a71-refs/heads/main@{#992738}
---
 .../ui/views/close_bubble_on_tab_activation_helper.cc  | 10 +++++++++-
 .../ui/views/close_bubble_on_tab_activation_helper.h   |  1 +
 2 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/chrome/browser/ui/views/close_bubble_on_tab_activation_helper.cc b/chrome/browser/ui/views/close_bubble_on_tab_activation_helper.cc
index 79c9ee559764..e5833968c1bc 100644
--- a/chrome/browser/ui/views/close_bubble_on_tab_activation_helper.cc
+++ b/chrome/browser/ui/views/close_bubble_on_tab_activation_helper.cc
@@ -18,7 +18,8 @@ CloseBubbleOnTabActivationHelper::CloseBubbleOnTabActivationHelper(
 }
 
 CloseBubbleOnTabActivationHelper::~CloseBubbleOnTabActivationHelper() {
-  browser_->tab_strip_model()->RemoveObserver(this);
+  if (browser_)
+    browser_->tab_strip_model()->RemoveObserver(this);
 }
 
 void CloseBubbleOnTabActivationHelper::OnTabStripModelChanged(
@@ -35,3 +36,10 @@ void CloseBubbleOnTabActivationHelper::OnTabStripModelChanged(
     owner_bubble_ = nullptr;
   }
 }
+
+void CloseBubbleOnTabActivationHelper::OnTabStripModelDestroyed(
+    TabStripModel* tab_strip_model) {
+  DCHECK(browser_);
+  browser_->tab_strip_model()->RemoveObserver(this);
+  browser_ = nullptr;
+}
diff --git a/chrome/browser/ui/views/close_bubble_on_tab_activation_helper.h b/chrome/browser/ui/views/close_bubble_on_tab_activation_helper.h
index 0fca6d080d70..c72e3191d7a6 100644
--- a/chrome/browser/ui/views/close_bubble_on_tab_activation_helper.h
+++ b/chrome/browser/ui/views/close_bubble_on_tab_activation_helper.h
@@ -37,6 +37,7 @@ class CloseBubbleOnTabActivationHelper : public TabStripModelObserver {
       TabStripModel* tab_strip_model,
       const TabStripModelChange& change,
       const TabStripSelectionChange& selection) override;
+  void OnTabStripModelDestroyed(TabStripModel* tab_strip_model) override;
 
  private:
   raw_ptr<views::BubbleDialogDelegateView> owner_bubble_;  // weak, owns me.
-- 
2.38.1

